lakefile.lean

```diff
@@ -88,40 +73,65 @@ def getGithubBaseUrl (url : String) : Option String :=
   else
     .none
 
-def getGithubUrl (mod : Module) : IO String := do
-  let url ← getGitRemoteUrl mod.pkg.dir "origin"
-  let .some baseUrl := getGithubBaseUrl url
-      | throw <| IO.userError <|
-        s!"Could not interpret Git remote uri {url} as a Github source repo.\n"
-          ++ "See README on source URIs for more details."
-  let commit ← getProjectCommit mod.pkg.dir
-  let srcDir := filteredPath mod.pkg.config.srcDir
-  let pkgUri := "/".intercalate <| baseUrl :: "blob" :: commit :: srcDir
-  return appendLibModPath pkgUri '/' mod
+inductive UriSource
+  | github
+  | vscode
+  | file
 
-/--
-Return a File uri for the module.
--/
-def getFileUri (mod : Module) : IO String := do
-  let dir ← IO.FS.realPath (mod.pkg.dir / mod.pkg.config.srcDir)
-  return appendLibModPath s!"file://{dir}" FilePath.pathSeparator mod
+instance : ComputeHash UriSource Id := ⟨fun
+  | .github => .mk 0
+  | .vscode => .mk 1
+  | .file => .mk 2⟩
 
-/--
-Return a VSCode uri for the module.
--/
-def getVSCodeUri (mod : Module) : IO String := do
-  let dir ← IO.FS.realPath (mod.pkg.dir / mod.pkg.config.srcDir)
-  return appendLibModPath s!"vscode://file/{dir}" FilePath.pathSeparator mod
+def UriSource.parse : LogIO UriSource := do
+  match ← IO.getEnv "DOCGEN_SRC" with
+  | .none | .some "github" | .some "" => pure .github
+  | "vscode" => pure .vscode
+  | "file" => pure .file
+  | _ => error "$DOCGEN_SRC should be github, file, or vscode."
 
-/--
-Attempt to determine URI to use for the module source file.
--/
-def getSrcUri (mod : Module) : IO String := do
-  match ←IO.getEnv "DOCGEN_SRC" with
-  | .none | .some "github" | .some "" => getGithubUrl mod
-  | .some "vscode" => getVSCodeUri mod
-  | .some "file" => getFileUri mod
-  | .some _ => throw <| IO.userError "$DOCGEN_SRC should be github, file, or vscode."
+-- try to make lake rebuild if the env var chaanges
+target env.DOCGEN_SRC : UriSource := do
+  let src ← UriSource.parse
+  addTrace (pureHash src)
```

Comment:
@tydeu, just to check I understand: this is pointless and does not cause the build to be invalidated when the environment variable changes?

Reply:
Correct. In fact, this literally does nothing as this call is not even within `JobM` (the lift from `JobM` of the call to the `FetchM` of the target will discard the trace state).

Reply:
Is it meaningful to put this within `JobM`?

Reply:
@eric-wieser Yes, it would be meaningful inside `JobM`. One way to do this is via `Job.async`. Another is by constructing the `Job` with the the proper trace (e.g., via `ofTask`). Unfortunately, the API here is not as polished as it would ideally be. That is something I hope to fix in the future.